### PR TITLE
add support for user and system labels

### PR DESCRIPTION
**Pull request by ErezLevip**
## Summary by CodeRabbit

* **New Features**
  * Optional inclusion of system labels in reported metrics with configurable user-label override, deterministic ordering and de-duplication.

* **Tests**
  * Extensive unit, integration, edge-case, and benchmark tests covering label composition, precedence, ordering, special-character handling, and performance.

* **Chores**
  * Added JSON-parsing dependency and related test dependencies.

The flags will be enabled using
```
    EnableSystemLabels:    true,
    UserLabelsOverride:    true,
```

User Metrics will always be added, if UserLabelsOverride flag is enabled it will override duplicates.
System Labels will added only if EnableSystemLabels is enabled and will not override

### CODE REVIEW

File: collectors/monitoring_collector.go [hunk 1]
```diff
@@ -574,3 +596,49 @@ func (c *MonitoringCollector) keyExists(labelKeys []string, key string) bool {
...
 	return false
 }
+
+func (c *MonitoringCollector) addSystemLabels(raw googleapi.RawMessage, labelKeys *[]string, labelValues *[]string) {
+	// Early exit for empty, null, or invalid JSON
+	if len(raw) == 0 {
+		return
+	}
+
+	result := gjson.ParseBytes(raw)
```
Comment (yosizelensky) [inline collectors/monitoring_collector.go:684]: why gjson?

File: collectors/monitoring_collector.go [hunk 2]
```diff
@@ -574,3 +596,49 @@ func (c *MonitoringCollector) keyExists(labelKeys []string, key string) bool {
...
+	result.ForEach(func(key, value gjson.Result) bool {
+		if !c.keyExists(*labelKeys, key.String()) {
+			*labelKeys = append(*labelKeys, key.String())
+			*labelValues = append(*labelValues, value.String())
+		}
+		return true // continue iteration
+	})
+}
+
+func (c *MonitoringCollector) addOrOverrideLabels(labelKeys *[]string, labelValues *[]string, key string, value string, override bool) {
```
Comment (yosizelensky) [inline collectors/monitoring_collector.go:700]: are you sure we need ptr to slice?

File: collectors/monitoring_collector.go [hunk 3]
```diff
@@ -84,6 +88,10 @@ type MonitoringCollectorOptions struct {
 	DescriptorCacheTTL time.Duration
 	// DescriptorCacheOnlyGoogle decides whether only google specific descriptors should be cached or all
 	DescriptorCacheOnlyGoogle bool
+	// EnableSystemLabels decides if system labels from metadata should be added to metrics
+	EnableSystemLabels bool
+	// UserLabelsOverride decides if user labels should override system labels when there are conflicts
```
Comment (yosizelensky) [inline collectors/monitoring_collector.go]: override system labels or metric labels?

File: collectors/monitoring_collector.go [hunk 4]
```diff
@@ -574,3 +596,49 @@ func (c *MonitoringCollector) keyExists(labelKeys []string, key string) bool {
...
+
+	if !override {
+		return
+	}
+
+	// Override the value
+	(*labelValues)[c.findKeyIndex(*labelKeys, key)] = value
+}
+
+func (c *MonitoringCollector) findKeyIndex(labelKeys []string, key string) int {
```
Comment (yosizelensky) [inline collectors/monitoring_collector.go:715]: can't we build a map for fast lookup?

**Review by yosizelensky — CHANGES_REQUESTED** [review]

File: collectors/monitoring_collector.go (hunk 4 above)
Comment (ErezLevip) [reply collectors/monitoring_collector.go:715]: Redundant, allocating will be more expensive then iterating on a handful of labels

File: collectors/monitoring_collector.go (hunk 1 above)
Comment (ErezLevip) [reply collectors/monitoring_collector.go:684]: It's faster

File: collectors/monitoring_collector.go (hunk 2 above)
Comment (ErezLevip) [reply collectors/monitoring_collector.go:700]: yes

File: collectors/monitoring_collector.go (hunk 4 above)
Comment (ErezLevip) [reply collectors/monitoring_collector.go:715]: this code can parse millions of metrics

**Review by yosizelensky — APPROVED** [review]

**Review by yosizelensky — APPROVED** [review]
